File: services/modal-api/periogt_runtime/schemas.py
"""Pydantic request/response models for the PerioGT API."""

from __future__ import annotations

from pydantic import BaseModel, Field


# --- Request Models ---


class PredictRequest(BaseModel):
    smiles: str = Field(..., min_length=1, max_length=2000, description="Polymer repeat-unit SMILES with * connection points")
    property: str = Field(..., min_length=1, max_length=64, description="Property ID (e.g., 'eps', 'tg', 'density')")
    return_embedding: bool = Field(default=False, description="Also return the graph embedding vector")


class EmbeddingRequest(BaseModel):
    smiles: str = Field(..., min_length=1, max_length=2000, description="Polymer repeat-unit SMILES with * connection points")


class BatchPredictRequest(BaseModel):
    items: list[PredictRequest] = Field(..., max_length=100, description="List of prediction requests (max 100)")


# --- Response Models ---


class PredictionValue(BaseModel):
    value: float
    units: str


class ModelInfo(BaseModel):
    name: str = "PerioGT"
    checkpoint: str


class PredictResponse(BaseModel):
    smiles: str
    property: str
    prediction: PredictionValue
    embedding: list[float] | None = None
    model: ModelInfo
    request_id: str


class EmbeddingResponse(BaseModel):
    smiles: str
    embedding: list[float]
    dim: int
    request_id: str


class BatchPredictResponse(BaseModel):
    results: list[PredictResponse | ErrorDetail]
    request_id: str


class PropertyInfo(BaseModel):
    id: str
    label: str
    units: str


class PropertiesResponse(BaseModel):
    properties: list[PropertyInfo]


class HealthResponse(BaseModel):
    status: str
    model_loaded: bool
    checkpoints_present: bool
    gpu_available: bool


# --- Error Models ---


class ErrorDetail(BaseModel):
    code: str
    message: str
    details: dict | list | str | None = None


class ErrorResponse(BaseModel):
    error: ErrorDetail
    request_id: str

File: apps/web/lib/schemas.ts
import { z } from "zod/v4";

// --- Request schemas ---

export const smilesSchema = z
  .string()
  .transform((s) => s.trim())
  .pipe(
    z
      .string()
      .min(1, "SMILES is required")
      .max(2000, "SMILES too long (max 2000 characters)")
      .refine((s) => s.includes("*"), {
        message: "Polymer SMILES must include '*' connection points",
      }),
  );

export const predictRequestSchema = z.object({
  smiles: smilesSchema,
  property: z.string().min(1, "Property is required"),
  return_embedding: z.boolean().default(false),
});

export const embeddingRequestSchema = z.object({
  smiles: smilesSchema,
});

export const batchPredictRequestSchema = z.object({
  items: z
    .array(predictRequestSchema)
    .min(1, "At least one item required")
    .max(100, "Maximum 100 items per batch"),
});

// --- Response schemas ---

export const predictionValueSchema = z.object({
  value: z.number(),
  units: z.string(),
});

export const modelInfoSchema = z.object({
  name: z.string(),
  checkpoint: z.string(),
});

export const predictResponseSchema = z.object({
  smiles: z.string(),
  property: z.string(),
  prediction: predictionValueSchema,
  embedding: z.array(z.number()).nullable().optional(),
  model: modelInfoSchema,
  request_id: z.string(),
});

export const embeddingResponseSchema = z.object({
  smiles: z.string(),
  embedding: z.array(z.number()),
  dim: z.number(),
  request_id: z.string(),
});

export const errorDetailSchema = z.object({
  code: z.string(),
  message: z.string(),
  details: z.unknown().optional(),
});

export const batchPredictResponseSchema = z.object({
  results: z.array(z.union([predictResponseSchema, errorDetailSchema])),
  request_id: z.string(),
});

export const propertyInfoSchema = z.object({
  id: z.string(),
  label: z.string(),
  units: z.string(),
});

export const propertiesResponseSchema = z.object({
  properties: z.array(propertyInfoSchema),
});

export const healthResponseSchema = z.object({
  status: z.string(),
  model_loaded: z.boolean(),
  checkpoints_present: z.boolean(),
  gpu_available: z.boolean(),
});

// --- Inferred types ---

export type PredictRequest = z.infer<typeof predictRequestSchema>;
export type PredictResponse = z.infer<typeof predictResponseSchema>;
export type EmbeddingRequest = z.infer<typeof embeddingRequestSchema>;
export type EmbeddingResponse = z.infer<typeof embeddingResponseSchema>;
export type BatchPredictRequest = z.infer<typeof batchPredictRequestSchema>;
export type BatchPredictResponse = z.infer<typeof batchPredictResponseSchema>;
export type PropertyInfo = z.infer<typeof propertyInfoSchema>;
export type PropertiesResponse = z.infer<typeof propertiesResponseSchema>;
export type HealthResponse = z.infer<typeof healthResponseSchema>;
export type ErrorDetail = z.infer<typeof errorDetailSchema>;
